src/main/java/co/cask/gcp/datastore/sink/RecordToEntityTransformer.java

```diff
@@ -288,57 +288,78 @@ private Key transformKeyLiteralToKey(Schema.Field field, KeyFactory keyFactory,
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case BOOLEAN:
-        outValue = BooleanValue.newBuilder(record.get(fieldName))
+        return BooleanValue.newBuilder(record.get(fieldName))
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case BYTES:
         byte[] byteArray = record.get(fieldName);
-        outValue = BlobValue.newBuilder(Blob.copyFrom(byteArray))
+        return BlobValue.newBuilder(Blob.copyFrom(byteArray))
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
       case RECORD:
         StructuredRecord nestedRecord = record.get(fieldName);
         FullEntity.Builder<IncompleteKey> nestedBuilder = FullEntity.newBuilder();
         Objects.requireNonNull(fieldSchema.getFields(), "Nested Schema fields cannot be empty").forEach(
           nestedField -> nestedBuilder.set(nestedField.getName(),
                                            convertToValue(nestedField.getName(), nestedField.getSchema(),
                                                           nestedRecord, excludeFromIndex)));
-        outValue = EntityValue.newBuilder(nestedBuilder.build())
+        return EntityValue.newBuilder(nestedBuilder.build())
           .setExcludeFromIndexes(excludeFromIndex)
           .build();
-        break;
+      case ARRAY:
+        Schema elementSchema = Schema.recordOf("arrayElementSchema",
+                                               Schema.Field.of(fieldName, fieldSchema.getComponentSchema()
+                                               ));
+        List<?> arrayValues = record.get(fieldName);
+        List<Value<?>> values = arrayValues.stream()
+          .map(value -> convertToValue(fieldName, elementSchema.getField(fieldName).getSchema(),
+                                       StructuredRecord.builder(elementSchema)
+                                         .set(fieldName, value)
+                                         .build(), false))
+          .collect(Collectors.toList());
+
+        ListValue.Builder listValueBuilder = ListValue.newBuilder();
+        if (!values.isEmpty()) {
+          listValueBuilder.addValue(values.get(0), values.subList(1, values.size()).toArray(new Value[0]));
+        }
+        // According to Datastore rule: list cannot be excluded from indexes
+        return listValueBuilder.build();
       case UNION:
+        // simple UNION type
         if (fieldSchema.isNullable()) {
-          outValue = convertToValue(fieldName, fieldSchema.getNonNullable(), record, excludeFromIndex);
-          break;
+          return convertToValue(fieldName, fieldSchema.getNonNullable(), record, excludeFromIndex);
+        }
+        // complex UNION type
+        for (Schema unionSchema : fieldSchema.getUnionSchemas()) {
+          try {
+            return convertToValue(fieldName, unionSchema, record, excludeFromIndex);
+          } catch (Exception e) {
```

Comment:
Complex union contains several types that given value might be of. We try to convert value to given type until succeed, in case of exception, we ignore it and we move the next possibility. In the end if we exhausted all possibilities, we throw the exception.